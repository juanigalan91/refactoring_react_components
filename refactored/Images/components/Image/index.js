import React from 'react';
import PropTypes from 'prop-types';

const namespace = 'image';
const Image = (props) => {
    const {
        link,
        id,
        src,
        title
    } = props;

    return (
        <li className={namespace} key={id}>
            <a href={link} className={`${namespace}__link`} target="_blank" rel="noopener noreferrer">
                <img
                    src={src}
                    alt={title}
                    className={`${namespace}__img`}
                />
                <h3 className={`${namespace}__img_title`}>{title}</h3>
            </a>
        </li>
    );
};

Image.propTypes = {
    link: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    src: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
};

export default Image;
