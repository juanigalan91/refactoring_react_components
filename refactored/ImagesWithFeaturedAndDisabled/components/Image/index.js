import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const namespace = 'image';
const Image = (props) => {
    const {
        link,
        src,
        isAvailable,
        isFeatured,
        notAvailableDisclaimer,
        id,
        title
    } = props;

    const className = classNames(namespace, {
        [`${namespace}__featured`]: isFeatured,
        [`${namespace}__not-available`]: !isAvailable,
    });

    return (
        <li className={className} key={id}>
            <a
                href={link}
                className={`${namespace}__link`}
                target="_blank"
                rel="noopener noreferrer"
                alt={title}
            >
                {
                    !isAvailable
                    && <span className={`${namespace}__not-available-disclaimer`}>{notAvailableDisclaimer}</span>
                }
                <img
                    src={src}
                    alt={title}
                    className={`${namespace}__img`}
                />
                <h3 className={`${namespace}__img_title`}>{title}</h3>
            </a>
        </li>
    );
};

Image.propTypes = {
    link: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    isAvailable: PropTypes.bool.isRequired,
    isFeatured: PropTypes.bool.isRequired,
    notAvailableDisclaimer: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
};

export default Image;
