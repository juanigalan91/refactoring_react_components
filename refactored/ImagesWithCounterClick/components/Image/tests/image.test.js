import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Image from '..';
import { render, cleanup, fireEvent } from 'react-testing-library';

afterEach(() => {
    cleanup();
});

const image = {
    id: 1,
    title: 'Red Dead Redemption 2',
    src: 'https://m.media-amazon.com/images/I/81z80lFufgL._AC_UL872_FMwebp_QL65_.jpg',
    link: 'https://www.amazon.com/Red-Dead-Redemption-2-PlayStation-4/dp/B01M5DZ525/ref=sr_1_3?keywords=ps4+games&qid=1551459125&s=gateway&sr=8-3',
    isFeatured: true,
    isAvailable: true,
    notAvailableDisclaimer: 'Not Available',
    track: jest.fn()
};

const component = (
    <Image {...image} />
);

describe('Image component', () => {
    it('renders a simple image component', () => {
        const renderer = new ShallowRenderer();

        expect(renderer.render(component)).toMatchSnapshot();
    });

    it('invokes track function since the image is available', () => {
        const { container } = render(component);

        const link = container.querySelector('.image__link');
        fireEvent.click(link);

        expect(image.track).toBeCalled();
    });

    it('does not invoke track function since the image is not available', () => {
        const notCalledTrack = jest.fn();
        const { container } = render(<Image {...image} isAvailable={false} track={notCalledTrack} />);

        const link = container.querySelector('.image__link');
        fireEvent.click(link);

        expect(notCalledTrack).not.toBeCalled();
    });
});
