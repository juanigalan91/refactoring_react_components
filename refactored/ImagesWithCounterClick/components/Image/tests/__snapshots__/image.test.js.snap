// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component renders a simple image component 1`] = `
<li
  className="image image__featured"
>
  <a
    alt="Red Dead Redemption 2"
    className="image__link"
    href="https://www.amazon.com/Red-Dead-Redemption-2-PlayStation-4/dp/B01M5DZ525/ref=sr_1_3?keywords=ps4+games&qid=1551459125&s=gateway&sr=8-3"
    onClick={[Function]}
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Red Dead Redemption 2"
      className="image__img"
      src="https://m.media-amazon.com/images/I/81z80lFufgL._AC_UL872_FMwebp_QL65_.jpg"
    />
    <h3
      className="image__img_title"
    >
      Red Dead Redemption 2
    </h3>
  </a>
</li>
`;
